rpc-broker/src/rpc-broker.c

```diff
@@ -0,0 +1,444 @@
+    ws_context = create_ws_context(args->port);
+    if (dom0)
+        dbus_broker_policy = build_policy(args->rule_file);
+
+    if (!ws_context)
+        DBUS_BROKER_ERROR("WebSockets-Server");
+
+    DBUS_BROKER_EVENT("<WebSockets-Server has started listening> [Port: %d]",
+                        args->port);
+
+    while (dbus_broker_running) {
+
+        lws_service(ws_context, WS_LOOP_TIMEOUT);
+        service_ws_signals();
+
+        if (reload_policy) {
+            free_policy();
+            dbus_broker_policy = build_policy(args->rule_file);
+            reload_policy = false;
+        }
+    }
+
+    lws_ring_destroy(ring);
+    lws_context_destroy(ws_context);
+}
+
+static void close_connection(uv_handle_t *handle)
+{
+    struct raw_dbus_conn *rdconn = (struct raw_dbus_conn *) handle->data;
+    free(rdconn);
+}
+
+void service_rdconn_cb(uv_poll_t *handle, int status, int events)
+{
+    struct raw_dbus_conn *rdconn = (struct raw_dbus_conn *) handle->data;
+
+    if (events & UV_READABLE)
+        broker_message(rdconn);
+    else if (events & UV_DISCONNECT)
+        uv_close((uv_handle_t *) handle, close_connection);
+}
+
+void run_rawdbus(struct dbus_broker_args *args)
+{
+    struct dbus_broker_server *server = start_server(args->port);
+    DBUS_BROKER_EVENT("<Server has started listening> [Port: %d]", args->port);
+
+    int default_socket = server->dbus_socket;
+    if (dom0)
+        dbus_broker_policy = build_policy(args->rule_file);
+
+    fd_set server_set;
+    uv_loop_t loop;
+    uv_loop_init(&loop);
+
+    while (dbus_broker_running) {
+
+        FD_ZERO(&server_set);
+        FD_SET(default_socket, &server_set);
+
+        struct timeval tv = { .tv_sec=0, .tv_usec=DBUS_BROKER_CLIENT_TIMEOUT };
+        int ret = select(default_socket + 1, &server_set, NULL, NULL, &tv);
+
+        if (ret > 0) {
+	        socklen_t clilen = sizeof(server->peer);
+	        int client = accept(default_socket,
+                               (struct sockaddr *) &server->peer, &clilen);
+            if (args->verbose) {
+                DBUS_BROKER_EVENT("<Client> [Port: %d Addr: %d Client: %d]",
+                                    args->port, server->peer.sin_addr.s_addr,
+                                                                     client);
+            }
+
+            struct raw_dbus_conn *rdconn = malloc(sizeof *rdconn);
+            struct raw_dbus_conn *sdconn = malloc(sizeof *sdconn);
+
+            rdconn->server = connect_to_system_bus();
+            rdconn->client = client;
+            rdconn->client_domain = 0;
+    	    /*
+             * When using rpc-broker over V4V, we want to be able to
+             * firewall against domids. The V4V interposer stores the
+             * domid as follows:
+             * sin.sin_addr.s_addr = htonl ((uint32_t) peer->domain | 0x1000000);
+             * If we're not using V4V, just return 0 for the domid.
+             */
+#ifdef HAVE_V4V
```

Comment:
+1, but I don't like "--test", I like to think of running rpc-broker outside of OpenXT as a valid use case.

Reply:
Why not just use libv4v and not perpetuate the bad habit of the interposer?

Reply:
Yeah, '--test' isn't good.  '--v4v' or '--v4v-hack' would make sense since we are enabling a v4v hack.  But since OpenXT is the main consumer, it seems weird to have all the production instances add that.  It could try to runtime detect, but that seems icky.  But, yeah, a better name.

Reply:
I don't think the interposer is a bad habit, IMO it's the best asset of V4V!
Just like sshd for example, rpc-broker is a tcp/ip program that can be used over V4V via the interposer. I just has a bit of logic to add a functionality when running over V4V.
And if only for being able to run in travis-ci, I think it's worth it not to have a hard dep on V4V.

Reply:
Yep, because LD_PRELOAD is such a great asset when you want to [break a system](https://pen-testing.sans.org/blog/2017/12/06/go-to-the-head-of-the-class-ld-preload-for-the-win).

Reply:
No need to be sarcastic here... No need to educate me on basic linking principles either. Any security-oriented software engineer knows the power, and therefore the danger, of LD_PRELOAD. And yes, if an attacker manages to get a shell in dom0, they can do a whole lot with LD_PRELOAD. However, it is already assumed that as soon as dom0 is compromised, it's game-over for OpenXT security, LD_PRELOAD or not.
As long as dom0 isolation is preserved, I don't see any problem using LD_PRELOAD wherever we want. Arguing against it would be like arguing that UIVM should do https to dom0, or arguing that dbus is unsafe, no matter how you use it... :wink:

Reply:
Apparently I do because there is a reason that when the AT_SECURE flag is set on an elf binary the glibc loader will flush LD_PRELOAD (and other environment variables generally considered "harmful") from the environment and why SELinux always flushes the environment on domain transition. LD_PRELOAD has always been considered a hack that is generally unsafe and never a good thing to do. If we eliminate any use of LD_PRELOAD, then among other good practices, we can begin dropping the unnecessary noatsecure override from our SELinux policy.

The fact of that matter is that there is no reason why a daemon being written from scratch specifically for OpenXT should not be made to support multiple transports, in particular v4v (soon to be argo) along with other "traditional" transports.